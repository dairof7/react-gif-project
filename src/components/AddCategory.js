import React, { useState } from 'react'
import PropTypes from 'prop-types'


export const AddCategory = ({ setCategories }) => {
    const [inputValue, setinputValue] = useState('')
    const handleInputChange = (e) => {
        setinputValue(e.target.value)        
        //console.log(e.target.value)
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        //cuando uso una funcion useState puedo usar el valor del estado anterior categ en este caso
        if ( inputValue.trim().length > 2 ) {
            setCategories(categ => [ inputValue,...categ ] )
            setinputValue('')
        }

    }

    return (
        <form onSubmit={ handleSubmit }>
            <input
                type='text'
                value={ inputValue }
                onChange={ ( e ) => handleInputChange(e) }
            />
        </form>
        

    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
