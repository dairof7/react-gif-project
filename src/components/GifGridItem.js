import PropTypes from 'prop-types'
import React from 'react'

export const GifGridItem = ( {title, url} ) => {
    return (
        <div className='card animate__animated animate__backInDown'>
            <img src={ url } alt={ title }/>
            <p>{ title }</p>
        </div>
    )
}



GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}