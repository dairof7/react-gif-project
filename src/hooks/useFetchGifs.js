import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";


export const useFetchGifs = ( category ) => {

    const [state, setState] = useState({
            data: [],
            loading: true
    }) 

     //los efectos no pueden ser asyncs
        useEffect(() => {
            //como getGifts es una promesa entonces ....            
            getGifs( category )
                .then( imgs => {
                        setState({
                            data: imgs,
                            loading: false
                        })
            } )
        }, [ category ])


    return state;
}
