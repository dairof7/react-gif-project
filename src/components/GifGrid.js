import React from 'react'
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';


export const GifGrid = ( { category } ) => {
    
    //en la desestructuracion puedo renombrar data si quiero, en este caso images
    const { data:images, loading } = useFetchGifs( category );
    //recibe una funcion y un arreglo de dependencias
    return (
        <>
        <h3 className='animate__animated animate__fadeInLeft'> { category } </h3>
        {/* si esta cargando aparece, condicional corto con && */}
        {loading && <p>Loading...</p>}
        <div className='card-grid'>
                {
                    images.map( (img) => 
                    <GifGridItem 
                        key={img.id}
                        //de esta forma se envia el paquete completo, es decir el array de objetos
                        //img={img}
                        //de esta otra forma estamos enviando ya cada uno de los objetos fuera del array
                        {...img}
                    />
                    )
                }
        </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}