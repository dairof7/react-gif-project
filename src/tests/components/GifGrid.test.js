const { shallow } = require("enzyme")
import '@testing-library/jest-dom'
import React from 'react'
import { GifGrid } from '../../components/GifGrid'
import { useFetchGifs } from '../../hooks/useFetchGifs'
jest.mock('../../hooks/useFetchGifs')
describe('test <GifGrid />', () => {
  const category = 'samurai'

  test('should show correctly snapshot', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    })
    const wrapper = shallow(<GifGrid category={ category }/>)
    expect ( wrapper ).toMatchSnapshot();
  })

  test('should show items when images are uploaded with useFectchGifts', () => {
  
    const gifs = [{
      id: 'ABC',
      url: 'https://localhost/lo/qie/sea.jpg',
      title: 'lo que sea'
    },
    {
      id: 'ABCDE',
      url: 'https://localhost/lo/qie/sea.jpg',
      title: 'lo que sea'
    }]
    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false
    })
    const wrapper = shallow(<GifGrid category={ category }/>);
    //se verifica si el parrafo desaparecio (loading)
    expect( wrapper ) .toMatchSnapshot()
    expect( wrapper.find('p').exists() ).toBe(false);
    expect( wrapper.find('GifGridItem').length ).toBe( gifs.length )
  })
})