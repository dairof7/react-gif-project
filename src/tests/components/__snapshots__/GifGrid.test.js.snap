// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <GifGrid /> should show correctly snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInLeft"
  >
     
    samurai
     
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`test <GifGrid /> should show items when images are uploaded with useFectchGifts 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeInLeft"
  >
     
    samurai
     
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="lo que sea"
      url="https://localhost/lo/qie/sea.jpg"
    />
    <GifGridItem
      id="ABCDE"
      key="ABCDE"
      title="lo que sea"
      url="https://localhost/lo/qie/sea.jpg"
    />
  </div>
</Fragment>
`;
